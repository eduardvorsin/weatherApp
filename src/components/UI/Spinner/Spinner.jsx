import PropTypes from 'prop-types';
import classes from './Spinner.module.css';

export default function Spinner({
  className,
  size,
}) {
  const spinnerSize = `${size}px`;
  const ringSize = `${0.8 * size}px`;

  return (
    <div
      className={`${classes.spinner} ${className}`}
      aria-label="loading spinner"
      style={{ '--size': spinnerSize }}
      title="loading spinner"
    >
      <div
        className={classes.spinner__ring}
        style={{ '--ring-size': ringSize }}
      />
      <div
        className={classes.spinner__ring}
        style={{ '--ring-size': ringSize }}
      />
      <div
        className={classes.spinner__ring}
        style={{ '--ring-size': ringSize }}
      />
      <div
        className={classes.spinner__ring}
        style={{ '--ring-size': ringSize }}
      />
    </div>
  );
}

Spinner.defaultProps = {
  className: '',
  size: 100,
};

Spinner.propTypes = {
  className: PropTypes.string,
  size: PropTypes.number,
};
