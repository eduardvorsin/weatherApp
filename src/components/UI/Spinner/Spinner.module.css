.spinner {
  position: relative;
  width: var(--size);
  height: var(--size);
  margin: 0 auto;
}

.spinner__ring {
  width: var(--ring-size);
  height: var(--ring-size);
  position: absolute;
  margin: 8px;
  border: 8px solid #fff;
  border-radius: 50%;
  animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  border-color: #fff transparent transparent transparent;
}

.spinner__ring:nth-child(1) {
  animation-delay: -0.45s;
}

.spinner__ring:nth-child(2) {
  animation-delay: -0.3s;
}

.spinner__ring:nth-child(3) {
  animation-delay: -0.15s;
}

@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}
