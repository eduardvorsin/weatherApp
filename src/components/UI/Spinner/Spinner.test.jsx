import { render, screen } from '@testing-library/react';
import Spinner from './Spinner';

describe('Spinner component tests', () => {
  it('renders correctly', () => {
    render(<Spinner
      size={60}
      className="test-spinner"
    />);

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
  });

  it('renders correctly without props', () => {
    render(<Spinner />);

    expect(screen.getByTitle('loading spinner')).toBeInTheDocument();
  });

  it('the classname is correctly assigned to the element', () => {
    render(<Spinner
      className="test-spinner"
    />);

    expect(screen.getByTitle('loading spinner')).toHaveClass('test-spinner');
  });

  it('if the prop size is 150, then the width and height of the spinner are also 150', () => {
    render(<Spinner
      className="test-spinner"
      size={150}
    />);

    expect(screen.getByTitle('loading spinner')).toHaveStyle({
      width: '150',
      height: '150',
    });
  });

  it('snapshot with default size', () => {
    render(<Spinner />);

    expect(screen.getByTitle('loading spinner')).toMatchSnapshot();
  });

  it('snapshot with custom size', () => {
    render(<Spinner
      size={30}
    />);

    expect(screen.getByTitle('loading spinner')).toMatchSnapshot();
  });
});
