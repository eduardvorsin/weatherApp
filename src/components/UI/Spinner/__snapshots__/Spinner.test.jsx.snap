// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner component tests snapshot with custom size 1`] = `
<div
  aria-label="loading spinner"
  class="spinner "
  style="--size: 30px;"
  title="loading spinner"
>
  <div
    class="spinner__ring"
    style="--ring-size: 24px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 24px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 24px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 24px;"
  />
</div>
`;

exports[`Spinner component tests snapshot with default size 1`] = `
<div
  aria-label="loading spinner"
  class="spinner "
  style="--size: 100px;"
  title="loading spinner"
>
  <div
    class="spinner__ring"
    style="--ring-size: 80px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 80px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 80px;"
  />
  <div
    class="spinner__ring"
    style="--ring-size: 80px;"
  />
</div>
`;
